lib/db/index.js

```diff
@@ -10,68 +10,116 @@ const logger = require('../logging')('db');
 const klass = config.get('db.driver') === 'mysql' ?
   require('./mysql') : require('./memory');
 
-function loadProviders() {
+/**
+ * Ensure that the database contains references
+ * for all the avatar providers mentioned in config.
+ */
+function initializeProvidersFromConfig(store) {
   var providers = Object.keys(config.get('img.providers'));
   logger.debug('providers.from-config', { providers: providers });
-  return P.all(providers.map(function(name) {
-    return exports.getProviderByName(name).then(function(provider) {
+  return P.all(providers.map(name => {
+    return store.getProviderByName(name).then(provider => {
       if (provider) {
         logger.debug('providers.exists',  { name: name });
       } else {
-        return exports.addProvider(name);
+        return store.addProvider(name);
       }
     });
   }));
 }
 
-var driverPromise;
-var driver;
+let driverPromise;
+let driver;
+
+/**
+ * Return a promise of a database instance,
+ * using the existing instance if available
+ * but creating a new one if necessary.
+ */
 function withDriver() {
   if (driver) {
     return P.resolve(driver);
   }
   if (driverPromise) {
     return driverPromise;
   }
+  let connectPromise;
   if (config.get('db.driver') === 'mysql') {
-    driverPromise = klass.connect(config.get('mysql'));
+    connectPromise = klass.connect(config.get('mysql'));
   } else {
-    driverPromise = klass.connect();
+    connectPromise = klass.connect();
   }
-  return driverPromise.then(function(store) {
+  driverPromise = connectPromise.then(store => {
     logger.debug('connected', config.get('db.driver'));
-    driver = store;
-    driverPromise = null;
-  }).then(loadProviders).then(function() {
-    return driver;
+    return initializeProvidersFromConfig(store).then(() => {
+      driver = store;
+      driverPromise = null;
+      return driver;
+    });
   });
+  return driverPromise;
 }
 
+
+/**
+ * Shut down the active database instance, if any.
+ * Calls to the DB after this function has completed,
+ * will automatically create a fresh new instance.
+ */
+exports.finalize = function finalize() {
+  if (driver) {
+    const d = driver;
+    driver = null;
+    return d.disconnect();
+  }
+  if (driverPromise) {
+    const dPromise = driverPromise;
+    driverPromise = null;
+    return dPromise.then(d => {
+      return d.disconnect();
+    });
+  }
+  return P.resolve();
+};
+
+
+/**
+ * Clear database state and shutdown the active instance.
+ * This is used by tests during their teardown phase.
+ */
+exports._teardown = function _teardown() {
+  return exports._clear().then(() => {
+    return exports.finalize();
+  });
+};
```

Comment:
As a side-effect of moving to nyc, the test runner would hang after completing all the tests when run with `DB=mysql`.  It turns out it was blocking on open connections in the MySQL connection pool, so I've added some explicit cleanup logic in the tests.  I also added some explanatory comments in this file because it look me a while to remember what it was doing.